src/ccs/ir/circuit.rs

```diff
@@ -0,0 +1,579 @@
+use std::{collections::HashMap, hash::Hash};
+
+use super::{assignments::*, CoeffsForProds};
+use crate::{
+    ccs::compiler::step_selector::{SelectorsForALLMatrix, SelectorsForALLSteps, StepSelector},
+    field::Field,
+    util::UUID,
+};
+
+pub type MatrixCoeffsAndOffset<F> = Vec<Vec<(CoeffsForProds<F>, usize)>>;
+pub type SelectorsOffsetAndCoeff<F> = Vec<Vec<(usize, F)>>;
+
+// ((step_idx, step_UUID), assignment_offset)
+pub type AssignmentOffsets = HashMap<(usize, u128), usize>;
+
+pub fn hadamard_product<F: Field + From<u64> + Hash>(vec1: &[F], vec2: &[F]) -> Vec<F> {
+    assert_eq!(vec1.len(), vec2.len());
+    vec1.iter()
+        .zip(vec2.iter())
+        .map(|(&v1, &v2)| v1 * v2)
+        .collect()
+}
+
+pub fn vec_add<F: Field + From<u64> + Hash>(vec: &[Vec<F>]) -> Vec<F> {
+    assert!(vec.len() > 1);
+    vec.iter().fold(vec![F::ZERO; vec[0].len()], |acc, vec| {
+        acc.iter().zip(vec.iter()).map(|(&a, &v)| a + v).collect()
+    })
+}
+
+// input vector z = [1, x, w]
+pub struct Z<F: Field + From<u64>> {
```

Comment:
@10to4 is Z pub or pub(crate) or pub(super), if it is pub it needs more documentation and explanation of what it is.

Reply:
ok